internal/server/auth/method/github/server.go

```diff
@@ -0,0 +1,166 @@
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	githubUserAPI = "https://api.github.com/user"
+)
+
+// OAuth2Client is our abstraction of communication with an OAuth2 Provider.
+type OAuth2Client interface {
+	AuthCodeURL(state string, opts ...oauth2.AuthCodeOption) string
+	Exchange(ctx context.Context, code string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error)
+	Client(ctx context.Context, t *oauth2.Token) *http.Client
+}
+
+const (
+	storageMetadataGithubEmail = "io.flipt.auth.github.email"
+	storageMetadataGithubName  = "io.flipt.auth.github.name"
+)
+
+// Server is an Github server side handler.
+type Server struct {
+	logger       *zap.Logger
+	store        storageauth.Store
+	config       config.AuthenticationConfig
+	oauth2Config OAuth2Client
+
+	auth.UnimplementedAuthenticationMethodGithubServiceServer
+}
+
+// NewServer constructs a Server.
+func NewServer(
+	logger *zap.Logger,
+	store storageauth.Store,
+	config config.AuthenticationConfig,
+) *Server {
+	return &Server{
+		logger: logger,
+		store:  store,
+		config: config,
+		oauth2Config: &oauth2.Config{
+			ClientID:     config.Methods.Github.Method.ClientId,
+			ClientSecret: config.Methods.Github.Method.ClientSecret,
+			Endpoint:     oauth2GitHub.Endpoint,
+			RedirectURL:  callbackURL(config.Methods.Github.Method.RedirectAddress),
+			Scopes:       []string{strings.Join(config.Methods.Github.Method.Scopes, ":")},
+		},
+	}
+}
+
+// RegisterGRPC registers the server as an Server on the provided grpc server.
+func (s *Server) RegisterGRPC(server *grpc.Server) {
+	auth.RegisterAuthenticationMethodGithubServiceServer(server, s)
+}
+
+func callbackURL(host string) string {
+	// strip trailing slash from host
+	host = strings.TrimSuffix(host, "/")
+	return host + "/auth/v1/method/github/callback"
+}
+
+// AuthorizeURL will return a URL for the client to redirect to for completion of the OAuth flow with GitHub.
+func (s *Server) AuthorizeURL(ctx context.Context, req *auth.AuthorizeURLRequest) (*auth.AuthorizeURLResponse, error) {
+	u := s.oauth2Config.AuthCodeURL(req.State)
+
+	return &auth.AuthorizeURLResponse{
+		AuthorizeUrl: u,
+	}, nil
+}
+
+// Callback is the OAuth callback method for Github authentication. It will take in a Code
+// which is the OAuth grant passed in by the OAuth service, and exchange the grant with an Authentication
+// that includes the user information.
+func (s *Server) Callback(ctx context.Context, r *auth.CallbackRequest) (*auth.CallbackResponse, error) {
+	if r.State != "" {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return nil, errors.ErrUnauthenticatedf("missing state parameter")
+		}
+
+		state, ok := md["flipt_client_state"]
+		if !ok || len(state) < 1 {
+			return nil, errors.ErrUnauthenticatedf("missing state parameter")
+		}
+
+		if r.State != state[0] {
+			return nil, errors.ErrUnauthenticatedf("unexpected state parameter")
+		}
+	}
```

Comment:
Might be nice to DRY this up into a utility function.
It could operate over the callback request or this interface:
```go
type Stater interface {
    GetState() string
}
```

Something like `methods.CallbackValidateState()`

Reply:
Then we can share it between both callback functions.